### PR TITLE
Direct Exporter - Blob store

### PR DESCRIPTION
**Pull request by sachintendulkar576123**
## Proposed changes

This PR updates Direct Exporter to Azure

- Implemented the AzureBlobSpanExporter() to upload OpenTelemetry traces to Azure Blob Storage.
- Added necessary dependencies for Azure support in pyproject.toml
- Instructions for installing Azure Blob exporters in Monocle_User_Guide.md

## Types of changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/monocle2ai/monocle/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the CLA
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by oi-raanne — CHANGES_REQUESTED** [review]

File: src/monocle_apptrace/exporters/azure/azure_exporter.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+# Set up logging
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+class AzureBlobSpanExporter(SpanExporter):
+    def __init__(self, connection_string=None, container_name=None):
+        # Use default values if none are provided
+        if connection_string is None:
+            connection_string = os.getenv('AZURE_STORAGE_CONNECTION_STRING')
+            if connection_string is None:
```
Comment (kshitiz-okahu) [inline src/monocle_apptrace/exporters/azure/azure_exporter.py:30]: You can use 
if not connection_string:

if not container_name:


This checks for empty string as well

**Comment by kshitiz-okahu** [comment]
adding an example with usage of the class would be good

**Review by kshitiz-okahu — CHANGES_REQUESTED** [review]

File: src/monocle_apptrace/exporters/azure/azure_exporter.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                break  # Authentication errors should not be retried
+            except Exception as e:
+                logger.warning(f"Retry {attempt}/{MAX_RETRIES} failed due to: {str(e)}")
+                sleep_time = BACKOFF_FACTOR * (2 ** attempt) + random.uniform(0, 1)
+                logger.info(f"Waiting for {sleep_time:.2f} seconds before retrying...")
+                time.sleep(sleep_time)
+                attempt += 1
+
+        logger.error("Max retries exceeded. Failed to upload spans to Azure Blob Storage.")
+        raise ServiceRequestError(message="Azure Blob Upload Endpoint")
```
Comment (oi-raanne) [inline src/monocle_apptrace/exporters/azure/azure_exporter.py:133]: what does the message 'Azure Blob Upload Endpoint` signify ?

File: src/monocle_apptrace/exporters/azure/azure_exporter.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        batch_to_export = self.export_queue[:MAX_BATCH_SIZE]
+        serialized_data = self._serialize_spans(batch_to_export)
+        self.export_queue = self.export_queue[MAX_BATCH_SIZE:]  # Remove exported spans from the queue
+
+        try:
+            self._upload_to_blob_with_retry(serialized_data)
+        except Exception as e:
+            logger.error(f"Failed to upload span batch: {e}")
+
+    def _upload_to_blob_with_retry(self, span_data_batch: str):
```
[the remark this answers is not in the document]
Comment (oi-raanne) [reply src/monocle_apptrace/exporters/azure/azure_exporter.py:106]: `export` method does not support async

**Review by oi-raanne — CHANGES_REQUESTED** [review]

File: src/monocle_apptrace/exporters/azure/azure_exporter.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            except ServiceRequestError as e:
+                logger.warning(f"Network connectivity error: {e}. Retrying in {BACKOFF_FACTOR ** attempt} seconds...")
+                sleep_time = BACKOFF_FACTOR * (2 ** attempt) + random.uniform(0, 1)
+                time.sleep(sleep_time)
+                attempt += 1
+            except ClientAuthenticationError as e:
+                logger.error(f"Failed to authenticate with Azure Blob Storage: {str(e)}")
+                break  # Authentication errors should not be retried
+            except Exception as e:
+                logger.warning(f"Retry {attempt}/{MAX_RETRIES} failed due to: {str(e)}")
```
Comment (oi-raanne) [inline src/monocle_apptrace/exporters/azure/azure_exporter.py:126]: should we retry for any other exception?? We may not need to

**Review by oi-raanne — CHANGES_REQUESTED** [review]

File: src/monocle_apptrace/exporters/azure/azure_exporter.py (hunk 4 above)
Comment (oi-raanne) [reply src/monocle_apptrace/exporters/azure/azure_exporter.py]: should we retry in `Exception as e:` case ??

**Review by oi-raanne — CHANGES_REQUESTED** [review]

File: src/monocle_apptrace/exporters/span_exporter_base.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import time
```
Comment (oi-raanne) [inline src/monocle_apptrace/exporters/span_exporter_base.py:1]: the name of the file can be just 'base_exporter.py'

File: src/monocle_apptrace/exporters/azure/azure_exporter.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
+import os
```
Comment (oi-raanne) [inline src/monocle_apptrace/exporters/azure/azure_exporter.py:1]: the name of the this file can be just 'blob_exporter.py' or 'az_blob_exporter.py
'

**Review by oi-raanne — CHANGES_REQUESTED** [review]

File: src/monocle_apptrace/exporters/span_exporter_base.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import random
+import logging
+from abc import ABC, abstractmethod
+from azure.core.exceptions import ServiceRequestError, ClientAuthenticationError
+from opentelemetry.sdk.trace import ReadableSpan
+from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult
+from typing import Sequence
+import asyncio
```
Comment (oi-raanne) [inline src/monocle_apptrace/exporters/span_exporter_base.py:9]: is this used ?

**Review by oi-raanne — CHANGES_REQUESTED** [review]

File: src/monocle_apptrace/exporters/span_exporter_base.py (hunk 7 above)
Comment (sachintendulkar576123) [reply src/monocle_apptrace/exporters/span_exporter_base.py:9]: yes, we are using.

**Review by oi-raanne — APPROVED** [review]

**Comment by kshitiz-okahu** [comment]
Sync the branch with main

**Review by anshul7665 — APPROVED** [review]

**Review by kshitiz-okahu — APPROVED** [review]

**Pull request merged**